### PR TITLE
Add jax.scipy.special.boxcox and boxcox1p

### PR DESCRIPTION
**Pull request by Saumay**
Adds `jax.scipy.special.boxcox` and `jax.scipy.special.boxcox1p`. Both are common in stats / regression and neither has a JAX equivalent today.

The only subtlety is the `lmbda = 0` limit. Using the naive `(x**lmbda - 1) / lmbda` form catastrophically cancels as `lmbda * log(x) -> 0`, which is exactly the regime a Box-Cox sweep crosses. I went with the `expm1` rewrite instead:

```python
power_branch = jnp.expm1(safe_lmbda * jnp.log(x)) / safe_lmbda
```

ULP-level accurate through the limit; SciPy parity drops from `~1e-3` to `~1e-9` near `lmbda = 0`. The `lmbda == 0` arm still routes to plain `log` / `log1p` for the exact limit. `boxcox1p` gets the same treatment with `log1p` on the inner expression, which also kills the `1 + x` precision loss for tiny `x`.

### Tests

Both names registered in `JAX_SPECIAL_FUNCTION_RECORDS` with `test_autodiff=True`, so they pick up the existing parity-vs-SciPy sweep, jit check, and `check_grads`. `boxcox` uses `jtu.rand_positive`; `boxcox1p` uses `jtu.rand_uniform(low=-0.5, high=5.0)` to cover negative `x` without crossing `-1`. 20 cases pass locally. Zero-sized dim + jit verified.

Issue: scipy.special API parity (no specific issue, follows the ongoing scipy port).

### CODE REVIEW

File: jax/_src/scipy/special.py [hunk 1]
```diff
@@ -728,6 +728,68 @@ def entr(x: ArrayLike) -> Array:
...
+  Returns:
+    array containing Box-Cox transformed values.
+
+  See also:
+    :func:`jax.scipy.special.boxcox1p`
+  """
+  x, lmbda = promote_args_inexact("boxcox", x, lmbda)
+  lmbda_is_zero = lmbda == 0
+  safe_lmbda = jnp.where(lmbda_is_zero, jnp.ones_like(lmbda), lmbda)
+  power_branch = (jnp.power(x, safe_lmbda) - 1) / safe_lmbda
```
[the remark this answers is not in the document]
Comment (Saumay) [reply jax/_src/scipy/special.py]: Switched to `jnp.expm1(lmbda * jnp.log(x)) / lmbda` so the small-lmbda regime no longer loses precision through `power - 1`. Quick spot check against SciPy at `x = 1.5`: diff drops from `~1e-3` at `lmbda = 1e-5` (old impl) to `~1e-9` now, and matches `log(x)` at the exact `lmbda = 0` branch (in 50607d4c0).

Re: the gradient w.r.t. `lmbda` at `lmbda = 0`: agreed, `jnp.where` gives zero there instead of the true `0.5 * log(x)**2`. SciPy itself does not expose a gradient so there is no parity target, but a follow-up `custom_jvp` could patch the seam if anyone needs the analytic value. Happy to do it in a follow-up if useful.

Re: zero-sized dims: confirmed `boxcox(jnp.zeros((0,)), 0.5)` and `boxcox(jnp.ones((2, 0, 3)), 0.5)` round-trip cleanly under `jit`. No runtime branching on shape, so polymorphism is preserved.

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -728,6 +728,68 @@ def entr(x: ArrayLike) -> Array:
...
+  Returns:
+    array containing values of the shifted Box-Cox transform.
+
+  See also:
+    :func:`jax.scipy.special.boxcox`
+  """
+  x, lmbda = promote_args_inexact("boxcox1p", x, lmbda)
+  lmbda_is_zero = lmbda == 0
+  safe_lmbda = jnp.where(lmbda_is_zero, jnp.ones_like(lmbda), lmbda)
+  power_branch = (jnp.power(1 + x, safe_lmbda) - 1) / safe_lmbda
```
[the remark this answers is not in the document]
Comment (Saumay) [reply jax/_src/scipy/special.py]: Applied the same stability fix here: `jnp.expm1(lmbda * jnp.log1p(x)) / lmbda` instead of `(1 + x)**lmbda - 1`. That handles both the small-`x` precision loss in `1 + x` and the small-lmbda precision loss in `power - 1` in one shot (in 50607d4c0). Zero-sized dim shape polymorphism verified end-to-end through `jit`.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
Hi - it looks like there are some conflicting edits that prevent this from being merged. Could you please rebase these changes on main? Thanks!

**Comment by Saumay** [comment]
Rebased onto current main in 8e5dc0eb9. Conflicts resolved (clean replay).
